### PR TITLE
Add integration test for the request token endpoints

### LINKED ISSUE #16146: Need test for token request endpoint
So we do not get another P0 like #16078 / #16109

### PR DESCRIPTION
**Pull request by mrogers950**
fixes #16146

### CODE REVIEW

**Comment by mrogers950** [comment]
@enj @openshift/sig-security

File: test/integration/oauth_token_endpoint_test.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
+package integration
+
+import (
+	"crypto/tls"
+	"net/http"
+	"net/url"
+	"testing"
+
+	"golang.org/x/net/html"
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Er, nothing in std lib that does this?

File: test/integration/oauth_token_endpoint_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+func TestOAuthRequestTokenEndpoint(t *testing.T) {
+	// Set up server
+	masterOptions, err := testserver.DefaultMasterOptions()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	defer testserver.CleanupMasterEtcd(t, masterOptions)
+
+	masterOptions.OAuthConfig.IdentityProviders[0] = configapi.IdentityProvider{
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Why do you need this?

File: test/integration/oauth_token_endpoint_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	clientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	transport := knet.SetTransportDefaults(&http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: You should not need this, you have all the CA information available to you.

File: test/integration/oauth_token_endpoint_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		t.Fatalf("unexpected error %v", err)
+	}
+	defer tokenResp.Body.Close()
+
+	if tokenResp.StatusCode != http.StatusFound {
+		t.Fatalf("unexpected response code %v", tokenResp.StatusCode)
+	}
+
+	// Follow location header to auth
+	authRedirect, err := url.Parse(tokenResp.Header.Get("Location"))
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: I think you want to check to make sure you have something in the location header.

File: test/integration/oauth_token_endpoint_test.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	authReq, err := http.NewRequest("GET", authRedirect.String(), nil)
+	if err != nil {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	authReq.SetBasicAuth("foo", "bar")
+	authReq.Header.Set("Accept", "text/html; charset=UTF-8")
+	authResp, err := transport.RoundTrip(authReq)
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Maybe add a helper for:

Request -> Round trip -> Status Found -> Location

File: test/integration/oauth_token_endpoint_test.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+	if displayResp.StatusCode != http.StatusOK {
+		t.Fatalf("unexpected response code %v", displayResp.StatusCode)
+	}
+
+	// Parse the HTML body for the API token contained in a <code></code> block
+	tokenizer := html.NewTokenizer(displayResp.Body)
+
+	var apiToken string
+	for {
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: `for token := tokenizer.Next(); token != html.ErrorToken {`

File: test/integration/oauth_token_endpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+			break
+		}
+	}
+
+	if len(apiToken) == 0 {
+		t.Fatalf("no API token found in HTML body")
+	}
+
+	// Check the use of the bearer token.
+	masterURL.Path = "/oapi/v1/users/~"
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: I think once you have the token you should be able to build a client that uses it.  Then check to make sure the `whoami` call does not error and the user that it points to is correct.

**Comment by enj** [comment]
@liggitt none of this changes when there is an auth proxy in front of the master right?  Just making sure this covers the cases we care about.

**Comment by mrogers950** [comment]
@enj I addressed your comments (for now sticking to using golang.org/x/net/html as it's used in other integration tests)

File: test/integration/oauth_token_endpoint_test.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	clientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	anonConfig := restclient.Config{}
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: `anonConfig := restclient.AnonymousClientConfig(clientConfig)`

File: test/integration/oauth_token_endpoint_test.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+	resp, err := rt.RoundTrip(req)
+	if err != nil {
+		t.Fatalf("unexpected error %v", err)
+	}
+	if resp.StatusCode != expectedCode {
+		t.Fatalf("unexpected response code %v", resp.StatusCode)
+	}
+
+	return resp, resp.Header.Get("Location")
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: return location and the body's bytes and close the body in this function always.

File: test/integration/oauth_token_endpoint_test.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	var seenCode bool
+	for tokenType := tokenizer.Next(); tokenType != html.ErrorToken; {
+		token := tokenizer.Token()
+		if tokenType == html.StartTagToken && token.Data == "code" {
+			seenCode = true
+		}
+		if seenCode && tokenType == html.TextToken {
+			return token.Data
+		}
+		tokenType = tokenizer.Next()
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Move this into the `for .. {`

File: test/integration/oauth_token_endpoint_test.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	return ""
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Pass in `t` and fatal the test here instead.

File: test/integration/oauth_token_endpoint_test.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+	displayResp, _ := checkNewReqAndRoundTrip(t, transport, authHeaderLocation, false, http.StatusOK)
+	apiToken := getTokenFromDisplay(displayResp)
+	if apiToken == "" {
+		t.Fatalf("API Token not found in display")
+	}
+	displayResp.Body.Close()
+
+	// Verify use of the bearer token
+	userConfig := &anonConfig
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Make sure to copy the config and just get user `~`

```go
userConfig := *anonConfig
userConfig.BearerToken = token
userClient, err := userclient.NewForConfig(&userConfig)
if err != nil {
	t.Fatalf("Unexpected error: %v", err)
}

user, err := userClient.Users().Get("~", metav1.GetOptions{})
if err != nil {
	t.Fatalf("Unexpected error: %v", err)
}
if user.Name != "username" {
	t.Fatalf("Expected username as the user, got %v", user)
}
```

**Comment by mrogers950** [comment]
flake #15900

File: test/integration/oauth_token_endpoint_test.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+	if len(authHeaderLocation) == 0 {
+		t.Fatalf("no Location header")
+	}
+
+	displayResp, _ := checkNewReqAndRoundTrip(t, transport, authHeaderLocation, false, http.StatusOK)
+	apiToken := getTokenFromDisplay(t, displayResp)
+
+	// Verify use of the bearer token
+	userConfig := anonConfig
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Use `AnonymousClientConfig` again here - this is a pointer so the way you did it does not make a copy.

File: test/integration/oauth_token_endpoint_test.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	if user.Name != "foo" {
+		t.Fatalf("expected foo as the user, got %v", user.Name)
+	}
+}
+
+// Parse the HTML body for the API token contained in the first <code></code> block
+func getTokenFromDisplay(t *testing.T, body []byte) string {
+	b := ioutil.NopCloser(bytes.NewReader(body))
+	defer b.Close()
+	tokenizer := html.NewTokenizer(b)
```
Comment (enj) [inline test/integration/oauth_token_endpoint_test.go]: Can you not just do `tokenizer := html.NewTokenizer(bytes.NewReader(body))`?  Why do you need `ioutil.NopCloser`?

**Comment by mrogers950** [comment]
@enj fixed nits.

**Comment by enj** [comment]
This is approved because it only adds a test and does not change any existing behavior.

**Pull request merged**